### PR TITLE
iOS: Fixes #12968: Fix mobile app unable to attach file with special characters in the name

### LINKED ISSUE #12968: iPad app unable to attach file with a '#' in the name
### Operating system

iOS

### Joplin version

13.3.9

### Desktop version info

_No response_

### Current behaviour

Image

I created a PDF from an email that used the subject line to name the file, and the subject line had a hash symbol (#) in it. I had no control over the file name it produced.

I tried to import that into a Joplin note. It said the file did not exist.

### Expected behaviour

Properly import the file

### Logs

08-15T13:11:47: root: root.biometrics: biometricsDone <true>
08-15T13:11:47: root: root.biometrics: biometricsEnabled <false>
08-15T13:11:47: root: root.biometrics: shouldShowMainContent <true>
08-15T13:11:47: root: root.biometrics: this.state.sensorInfo {"enabled":false,"sensorsHaveChanged":false,"supportedSensors":""}
08-15T13:11:42: RevisionService: maintenance: Done in 105ms
08-15T13:11:42: RevisionService: collectRevisions: Created revisions for 0 notes
08-15T13:11:42: RevisionService: maintenance: Starting...
08-15T13:11:42: RevisionService: maintenance: Service is enabled
08-15T13:11:34: root: root.biometrics: biometricsDone <true>
08-15T13:11:34: root: root.biometrics: biometricsEnabled <false>
08-15T13:11:34: root: root.biometrics: shouldShowMainContent <true>
08-15T13:11:34: root: root.biometrics: this.state.sensorInfo {"enabled":false,"sensorsHaveChanged":false,"supportedSensors":""}
08-15T13:11:33: root: root.biometrics: biometricsDone <true>
08-15T13:11:33: root: root.biometrics: biometricsEnabled <false>
08-15T13:11:33: root: root.biometrics: shouldShowMainContent <true>
08-15T13:11:33: root: root.biometrics: this.state.sensorInfo {"enabled":false,"sensorsHaveChanged":false,"supportedSensors":""}
08-15T13:11:29: root: root.biometrics: biometricsDone <true>
08-15T13:11:29: root: root.biometrics: biometricsEnabled <false>
08-15T13:11:29: root: root.biometrics: shouldShowMainContent <true>
08-15T13:11:29: root: root.biometrics: this.state.sensorInfo {"enabled":false,"sensorsHaveChanged":false,"supportedSensors":""}
08-15T13:11:12: RevisionService: collectRevisions: Created revisions for 0 notes
08-15T13:11:12: RevisionService: maintenance: Done in 106ms
08-15T13:11:12: RevisionService: maintenance: Starting...
08-15T13:11:12: RevisionService: maintenance: Service is enabled
08-15T13:10:42: RevisionService: maintenance: Done in 114ms
08-15T13:10:42: RevisionService: collectRevisions: Created revisions for 0 notes
08-15T13:10:42: RevisionService: maintenance: Service is enabled
08-15T13:10:42: RevisionService: maintenance: Starting...
08-15T13:10:39: Could not attach file: Error: The file “Order Received - Order %23T3666673.pdf” couldn’t be opened because there is no such file.
Code: ENSCOCOAERRORDOMAIN260
Error: The file “Order Received - Order %23T3666673.pdf” couldn’t be opened because there is no such file.
    at promiseMethodWrapper (/Users/laurent/Library/Developer/Xcode/DerivedData/Joplin-aabednnutrlvgkdcsfhvjkegcpwu/Build/Intermediates.noindex/ArchiveIntermediates/Joplin/BuildProductsPath/Release-iphoneos/main.jsbundle:4960:45)
    at copyFile (/Users/laurent/Library/Developer/Xcode/DerivedData/Joplin-aabednnutrlvgkdcsfhvjkegcpwu/Build/Intermediates.noindex/ArchiveIntermediates/Joplin/BuildProductsPath/Release-iphoneos/main.jsbundle:457735:34)
    at ?anon_0_ (/Users/laurent/Library/Developer/Xcode/DerivedData/Joplin-aabednnutrlvgkdcsfhvjkegcpwu/Build/Intermediates.noindex/ArchiveIntermediates/Joplin/BuildProductsPath/Release-iphoneos/main.jsbundle:409201:60)
    at next (native)
    at asyncGeneratorStep (/Users/laurent/Library/Developer/Xcode/DerivedData/Joplin-aabednnutrlvgkdcsfhvjkegcpwu/Build/Intermediates.noindex/ArchiveIntermediates/Joplin/BuildProductsPath/Release-iphoneos/main.jsbundle:65327:19)
    at _next (/Users/laurent/Library/Developer/Xcode/DerivedData/Joplin-aabednnutrlvgkdcsfhvjkegcpwu/Build/Intermediates.noindex/ArchiveIntermediates/Joplin/BuildProductsPath/Release-iphoneos/main.jsbundle:65341:29)
    at tryCallOne (address at InternalBytecode.js:1:1180)
    at anonymous (address at InternalBytecode.js:1:1874)
    at apply (native)
    at anonymous (/Users/laurent/Library/Developer/Xcode/DerivedData/Joplin-aabednnutrlvgkdcsfhvjkegcpwu/Build/Intermediates.noindex/ArchiveIntermediates/Joplin/BuildProductsPath/Release-iphoneos/main.jsbundle:14625:26)
    at _callTimer (/Users/laurent/Library/Developer/Xcode/DerivedData/Joplin-aabednnutrlvgkdcsfhvjkegcpwu/Build/Intermediates.noindex/ArchiveIntermediates/Joplin/BuildProductsPath/Release-iphoneos/main.jsbundle:14514:17)
    at _callReactNativeMicrotasksPass (/Users/laurent/Library/Developer/Xcode/DerivedData/Joplin-aabednnutrlvgkdcsfhvjkegcpwu/Build/Intermediates.noindex/ArchiveIntermediates/Joplin/BuildProductsPath/Release-iphoneos/main.jsbundle:14552:17)
    at callReactNativeMicrotasks (/Users/laurent/Library/Developer/Xcode/DerivedData/Joplin-aabednnutrlvgkdcsfhvjkegcpwu/Build/Intermediates.noindex/ArchiveIntermediates/Joplin/BuildProductsPath/Release-iphoneos/main.jsbundle:14755:44)
    at __callReactNativeMicrotasks (/Users/laurent/Library/Developer/Xcode/DerivedData/Joplin-aabednnutrlvgkdcsfhvjkegcpwu/Build/Intermediates.noindex/ArchiveIntermediates/Joplin/BuildProductsPath/Release-iphoneos/main.jsbundle:3583:48)
    at anonymous (/Users/laurent/Library/Developer/Xcode/DerivedData/Joplin-aabednnutrlvgkdcsfhvjkegcpwu/Build/Intermediates.noindex/ArchiveIntermediates/Joplin/BuildProductsPath/Release-iphoneos/main.jsbundle:3447:45)
    at __guard (/Users/laurent/Library/Developer/Xcode/DerivedData/Joplin-aabednnutrlvgkdcsfhvjkegcpwu/Build/Intermediates.noindex/ArchiveIntermediates/Joplin/BuildProductsPath/Release-iphoneos/main.jsbundle:3555:15)
    at flushedQueue (/Users/laurent/Library/Developer/Xcode/DerivedData/Joplin-aabednnutrlvgkdcsfhvjkegcpwu/Build/Intermediates.noindex/ArchiveIntermediates/Joplin/BuildProductsPath/Release-iphoneos/main.jsbundle:3446:21)
    at invokeCallbackAndReturnFlushedQueue (/Users/laurent/Library/Developer/Xcode/DerivedData/Joplin-aabednnutrlvgkdcsfhvjkegcpwu/Build/Intermediates.noindex/ArchiveIntermediates/Joplin/BuildProductsPath/Release-iphoneos/main.jsbundle:3440:33)
08-15T13:10:39: Got type: application/pdf
08-15T13:10:39: Got file: file:///private/var/mobile/Containers/Data/Application/6081B7BA-BEE2-4E50-8517-502E7867A254/tmp/net.cozic.joplin-Inbox/Order Received - Order %23T3666673.pdf
08-15T13:10:12: RevisionService: collectRevisions: Created revisions for 1 notes
08-15T13:10:12: RevisionService: maintenance: Done in 138ms
08-15T13:10:12: RevisionService: maintenance: Service is enabled
08-15T13:10:12: RevisionService: maintenance: Starting...
08-15T13:09:48: checkDisabledSyncItemsNotification: No errors: Hiding notification
08-15T13:09:48: root: root.biometrics: biometricsDone <true>
08-15T13:09:48: root: root.biometrics: biometricsEnabled <false>
08-15T13:09:48: root: root.biometrics: shouldShowMainContent <true>
08-15T13:09:48: root: root.biometrics: this.state.sensorInfo {"enabled":false,"sensorsHaveChanged":false,"supportedSensors":""}

### PR DESCRIPTION
**Pull request by gherardi**
Fixes #12968 

**Problem**
Attaching or importing a file in iOS fails when the filename contains the # character. The application returns an error stating that the file does not exist.

From the logs reported in the related issue, the original file name is:
`Order Received - Order #T3666673.pdf`
However, the error logs show the application attempting to access:
`Order Received - Order %23T3666673.pdf`

The `#` character has a special meaning in URIs (it denotes fragments). When iOS provides the file to the application it automatically performs URL encoding and converts `#` into `%23`.

Later, when the application attempts to read or copy the file using this encoded path, the file system lookup fails because the actual file on disk still contains the # character. As a result, the file system cannot locate the file and throws a file not found error.

**Solution**
To resolve this issue without significantly altering the application logic, the file path should be normalized before interacting with the file system.

- **Typescript fix**: `decodeURI()` was replaced with `decodeURIComponent()` when handling file paths received from iOS. This ensures all percent-encoded characters are decoded before interacting with the filesystem.
- **iOS share extension**: File names are decoded using `stringByRemovingPercentEncoding`, source paths are normalized before copying files. This guarantees that files stored in the shared container always use their correct decoded filenames.
- A SharedResource interface was introduced to properly structure shared data.
- SharedData was updated to use typed resources instead of raw strings.
- A small path decoding utility was introduced to centralize URI normalization logic.

The fix now correctly supports filenames containing characters such as:
- `#` becomes `%23`
- `&` becomes `%26`
- `?` becomes `%3F`
- `%` becomes `%25`
- `spaces` becomes `%20`

**Test Plan**
- Added a test suite for the URI decoding utilities.
- 16 unit tests verify correct decoding behavior across multiple encoded characters.

**AI disclosure:**
A local LLM were used during the investigation and development process, specifically to help and analyse the issue logs and investigate the root cause related to URI encoding, it was also used to suggest test cases to validate the fix across different filename edge cases.

I have reviewed and understood all the changes, compiled the application locally, and tested the fix to confirm it works as intended.

### Before

https://github.com/user-attachments/assets/de420790-0651-4a94-b22b-78ef56b250b6

### After

https://github.com/user-attachments/assets/1e0d7b6d-7f90-473d-b220-7c1dbfb0a9d8

### CODE REVIEW

File: packages/app-mobile/utils/shareHandler.ts [hunk 1]
```diff
@@ -42,7 +43,24 @@ export default async (sharedData: SharedData, folderId: string, dispatch: Functi
...
+		if (decodedSharedData.resources && decodedSharedData.resources.length > 0) {
+			decodedSharedData.resources = decodedSharedData.resources.map(
+				(resource) => {
+					return {
+						...resource,
+						uri: decodeFileUri(resource.uri),
+					};
+				},
+			);
+		}
```
[the remark this answers is not in the document]
Comment (gherardi) [reply packages/app-mobile/utils/shareHandler.ts]: solved

File: packages/app-mobile/utils/pathDecoding.test.ts [hunk 2]
```diff
@@ -0,0 +1,99 @@
...
+				'file:///private/var/mobile/Containers/Shared Data/Order #123 & Tracking.pdf',
+			);
+		});
+
+		it('should handle already decoded URIs', () => {
+			const uri = 'file:///var/mobile/Order #T3666673.pdf';
+			const decoded = decodeFileUri(uri);
+			expect(decoded).toBe(uri);
+		});
+	});
```
[the remark this answers is not in the document]
Comment (gherardi) [reply packages/app-mobile/utils/pathDecoding.test.ts]: solved

File: packages/app-mobile/ios/ShareExtension/Source/ShareExtension/ShareViewController.m [hunk 3]
```diff
@@ -102,8 +102,13 @@ - (NSURL*)copyImageToSharedContainer:(UIImage*)image {
...
 - (NSURL*)copyURLToSharedContainer:(NSURL*)url {
   NSString* name = [url lastPathComponent];
+  // Decode the file name to handle URL-encoded characters (e.g., %23 for #)
+  // This must be done before creating the destination path
+  name = [name stringByRemovingPercentEncoding];
   NSURL* path = [self.sharedContainerURL URLByAppendingPathComponent:name];
-  [[NSFileManager defaultManager] copyItemAtPath:[url path] toPath:[path path] error:nil];
+  // Decode the URL path to handle special characters like # that are URL-encoded as %23
+  NSString* decodedSourcePath = [[url path] stringByRemovingPercentEncoding];
+  [[NSFileManager defaultManager] copyItemAtPath:decodedSourcePath toPath:[path path] error:nil];
```
[the remark this answers is not in the document]
Comment (gherardi) [reply packages/app-mobile/ios/ShareExtension/Source/ShareExtension/ShareViewController.m]: solved

**Comment by personalizedrefrigerator** [comment]
> Existing share extension integration continues to work correctly.

Thank you for the screen recordings! Since this also adjusts the native share extension code (for sharing data from another app to Joplin), please attach a screen recording that demonstrates that sharing a resource from another app with Joplin still works.

It might also make sense to split the changes to the native share extension code into a separate pull request.

**Comment by gherardi** [comment]
Thanks for the advice. I'll attach the screen recording of the share extension in a separate pull request, splitting up the changes.

**Comment by gherardi** [comment]
Unfortunately, the option to share with Joplin doesn't appear even when I'm on dev main branch (so without my changes). It might be an issue with my simulator, or maybe I need to configure something

https://github.com/user-attachments/assets/bd722a0b-01ce-4dbe-8085-379bcde9e207

**Comment by personalizedrefrigerator** [comment]
> It might be an issue with my simulator, or maybe I need to configure something

I think it's necessary to run Joplin using a different configuration. Just to the left of the "iPad Pro 11 inch (M5)" should be a dropdown that allows selecting different run configurations. It should be possible to select a "share extension" run configuration from that dropdown.

Also, the original issue (https://github.com/laurent22/joplin/issues/12968) doesn't seem very specific. Does the share extension needs to be updated to fix the issue? (Are the other changes sufficient?)

**Comment by gherardi** [comment]
No, it's not really necessary, the changes I made are enough to fix the bug, It would be great if you could do the merge

**Comment by laurent22** [comment]
Unfortunately we cannot merge if it hasn't been tested. Are the changes to the share extension necessary? And if so, why?

**Comment by gherardi** [comment]
> Are the changes to the share extension necessary?

I decided to remove the changes made to the share extension because they were not necessary to resolve this issue, now the changes in this commit only concern the way Joplin handles special characters

I re-tested the solution and everything works fine.

https://github.com/user-attachments/assets/b00aefcd-5c0a-4c3a-8152-4bf8aed843a7

**Comment by gherardi** [comment]
I understand that merging 8 commits for just a few lines of changes isn't great, if you ask I can close the pull request here and create another one with only the necessary changes

**Comment by personalizedrefrigerator** [comment]
> I understand that merging 8 commits for just a few lines of changes isn't great, if you ask I can close the pull request here and create another one with only the necessary changes

This isn't an issue. Commits will be [squashed](https://github.blog/open-source/git/squash-your-commits/#enter-commit-squashing) on merge.

**Comment by gherardi** [comment]
One final note, every time I want to test the mobile application in Xcode and need to install the POD dependencies with `pod install`, I encounter errors caused by outdated dependencies in the Podfile.lock (specifically `hermes-engine` and `React-Core-prebuilt`)
This forces me to update the lock file every time using the `pod update` command, which bumps some dependencies to their latest version.
Wouldn't it be better to permanently update that file?

File: packages/app-mobile/components/screens/ConfigScreen/NoteExportSection/NoteImportButton.tsx [hunk 4]
```diff
@@ -62,7 +62,7 @@ const NoteImportButton: FunctionComponent<Props> = props => {
 		const sourceFileUri = importFiles[0].uri;
 		const sourceFilePath = Platform.select({
 			default: sourceFileUri,
-			ios: decodeURI(sourceFileUri),
+			ios: decodeURIComponent(sourceFileUri),
```
Comment (laurent22) [inline packages/app-mobile/components/screens/ConfigScreen/NoteExportSection/NoteImportButton.tsx:65]: I don't quite understand this fix. Obviously decodeURI is for URIs, and this is a URI we are dealing with. So why does it need to be replaced with decodeURIComponent?

Comment (gherardi) [reply packages/app-mobile/components/screens/ConfigScreen/NoteExportSection/NoteImportButton.tsx:65]: While we are dealing with a URI, decodeURI is designed to ignore characters that have special meaning in a full web address (like `:`, `/`, `?`, `#`).

If the sourceFileUri contains a filename with a literal `#` or `&` that has been percent encoded, `decodeURI` will skip them, leaving the file path broken or incorrectly parsed.

`decodeURIComponent()` is intended for individual data parts. It decodes every percent encoded character, ensuring that special characters within the filename itself (like `image#1.jpg`) are restored correctly.

Comment (gherardi) [reply packages/app-mobile/components/screens/ConfigScreen/NoteExportSection/NoteImportButton.tsx:65]: `decodeURI` is designed to preserve URI delimiters like `#` and `?`. However, since our `sourceFileUri` represents a file path, those characters are actually part of the data (the filename) and not structural URL markers. Using `decodeURIComponent` ensures that filenames containing `#`, `&`, or `+` are fully restored so the OS can actually locate the file.

Comment (laurent22) [reply packages/app-mobile/components/screens/ConfigScreen/NoteExportSection/NoteImportButton.tsx:65]: That makes sense, thanks for clarifying!
